### PR TITLE
bug/1592_query_with_q_op_greater_or_equal

### PR DESCRIPTION
**Pull request by kzangeli**
Fixed issue #1592 
1592: return error for q with GTE without right-hand.
q with LTE without right-hand also fixed

Added two 'future functests', for issues 1587 and 1589, to be fully incorporated once the issue 1675 is fixed.

### CODE REVIEW

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 1]
```diff
@@ -1282,9 +1294,9 @@ static void addBsonFilter
   }
   else
   {
-    std::string details = std::string("unknown query operator: ") + opr;
+    std::string details = std::string("unknown query operator: '") + opr + "'";
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Is this `details` used in some way in the response payload to client? Note that `'` is a forbidden char....

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 2]
```diff
@@ -1092,7 +1092,7 @@ static bool matchFilter
 * Add a BSON filter based on a "q token".
 *
 */
-static void addBsonFilter
+static bool addBsonFilter
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: This solution sufices for the case of q expressions in the GET /v2/entities. However, q expressions in subscriptions will be also problematic. 

This fact support the idea of do filter parsing at early stages, both for GET /v2/entities and subscriptions, and have the data as objects (i.e. https://github.com/telefonicaid/fiware-orion/issues/1705)

NTC (just a thought)

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 3]
```diff
@@ -1282,9 +1294,9 @@ static void addBsonFilter
...
   else
   {
-    std::string details = std::string("unknown query operator: ") + opr;
+    std::string details = std::string("unknown query operator: '") + opr + "'";
     alarmMgr.badInput(clientIp, details);
-    pushBackFilter = false;
+    return false;
   }
 
   if (pushBackFilter)
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Given the early-return which formarly corresponds to `pushBackFilter = false`, this if could be removed, replaced by just: 

```
filters.push_back(f);
```

As side effect, the `pushBackFilter` variable is no longer needed and can be completely removed from this function.

File: src/lib/mongoBackend/MongoGlobal.cpp [hunk 4]
```diff
@@ -1647,7 +1665,16 @@ bool entitiesQuery
     }
     else if (sco->type == SCOPE_TYPE_SIMPLE_QUERY)
     {
-      qStringFilters(sco->value, filters);
+      if (qStringFilters(sco->value, filters) != true)
```
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Would be useful to add a `FIXME #1705` above this line to explain that this checking will be removed once that issue will be addressed?

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: As far as I know, it is not used. Might be used in the future though.
Change for /  ?

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: I only changed this part, I believe there are more instances of **pushBackFilter = false**.
I can fix this though if you want.

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Sure

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: I think there aren't more instances. If that is correct, please go ahead with the fix.

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 1 above)
Comment (fgalan) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Ok.

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: I was wrong, it was in only one place.
Fixed in next commit.

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 1 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Fixed in f2b6eb3

File: test/functionalTest/cases/1587_q_query_has_type/q_query_has_type.test [hunk 5]
```diff
@@ -0,0 +1,192 @@
+# Copyright 2016 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/1587_q_query_has_type/q_query_has_type.test]: Although I undesrtand that having the test is working as it is now due to some steps are commented, I would suggest to rename it to .pending1675.DISABLED in order to make explicit when this test could be "recovered".

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 3 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Fixed in f2b6eb3

File: src/lib/mongoBackend/MongoGlobal.cpp (hunk 4 above)
Comment (kzangeli) [inline src/lib/mongoBackend/MongoGlobal.cpp]: Fixed in f2b6eb3

File: test/functionalTest/cases/1587_q_query_has_type/q_query_has_type.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1587_q_query_has_type/q_query_has_type.test]: Fixed in b069933

**Comment by fgalan** [comment]
LGTM (passing the ball to @crbrox for final LGTM + merge + delete branch)

**Comment by crbrox** [comment]
LGTM

**Pull request merged**